core/src/main/java/com/linecorp/armeria/server/VirtualHostBuilder.java

```diff
@@ -599,6 +614,39 @@ public VirtualHostBuilder decorator(
         return decorator(delegate -> new FunctionalDecoratingService<>(delegate, decoratingServiceFunction));
     }
 
+    /**
+     * Decorates {@link Service}s whose {@link Route} matches the specified {@code decorator}.
+     *
+     * @param route the route being decorated
+     * @param decorator the {@link Function} that decorates {@link Service}
+     * @param <T> the type of the {@link Service} being decorated
+     * @param <R> the type of the {@link Service} {@code decorator} will produce
+     */
+    public <T extends Service<HttpRequest, HttpResponse>, R extends Service<HttpRequest, HttpResponse>>
+    VirtualHostBuilder decorator(Route route, Function<T, R> decorator) {
```

Comment:
Oops I missed these methods. Do you think we need e.g. `decoratingRouter` if we have these? It's a whole new, fairly large public API, but is so similar to this one I'm not totally sure.

Reply:
Perhaps we have too many ways to bind things. :) How about removing this and sticking to the fluent API?

Reply:
At second thought, I feel like we need them. See https://github.com/line/armeria/pull/2040#discussion_r322638589